### PR TITLE
WIP: Well-kown URI

### PR DESCRIPTION
**Pull request by schulztr**
Using /.well-known/core for discovery

### ToDos
- [x] check that first parameter in link-format is always a link
- [x] check if link is absolute or relative

### CODE REVIEW

File: nodes/wot-discovery.js [hunk 1]
```diff
@@ -111,8 +112,9 @@ module.exports = function (RED) {
         }
 
         function _sendCoapDiscovery(address) {
+            node.log(wellKnownURI);
```
Comment (JKRhb) [inline nodes/wot-discovery.js:115]:
```suggestion
```

File: nodes/wot-discovery.js [hunk 2]
```diff
@@ -147,6 +159,39 @@ module.exports = function (RED) {
...
+                        links[i]=links[i].replace('<', '');
+                        links[i]=links[i].replace('>', '');
+                        node.log(links[i]);
+                        coapAddresses.forEach((address)=>{
+                            _sendCoapDiscovery(address, links[i]);
+                        });
+                    }
+                }
+            });
+        }
```
Comment (JKRhb) [inline nodes/wot-discovery.js:194]: The following is a rather complex suggestion which does both some refactoring and adds filter capabilities to the function. I already tested it with an ESP with WoT support and it worked :)

Maybe you could have a look on the TODO? A regEx assertion like you did before could be nice to ensure that the first parameter has the appropriate link format.

One thing we could discuss (and also refactor) is the actual filtering by resource and content type. In theory a resource could have multiple resource types which is why I was wondering if we should also convert the args to an array, iterate over them, and check if `"wot.thing"` is one of them. However, I guess a thing description should only have this one resource type which should make this additional step unnecessary. 

```suggestion
                 if (res.headers["Content-Format"] === "application/link-format") {
                    let links = data.toString().split(',');
                    
                    links = links.map(link => {
                        return link.split(";");
                    });
                    
                    links.forEach(link => {

                        let correctResourceType = false;
                        let correctContentType = false;
                        let path;

                        link.forEach(function (curentValue, index) {
                            if (index === 0) {
                                // First parameter must be the path
                                // TODO: Add assertion for </ ... > format
                                path = curentValue.substring(2, curentValue.length - 1);
                                return;
                            } else if (!path) {
                                return;
                            }

                            curentValue = curentValue.split("=");
                            let parameter = curentValue[0];
                            let values = curentValue[1]

                            switch (parameter) {
                                case "ct":
                                    if (values === "432") {
                                        correctContentType = true;
                                    }
                                    break;
                                case "rt":
                                    if (values === '"wot.thing"') {
                                        correctResourceType = true;
                                    }
                                    break;
                            }
                        })

                        if (correctContentType && correctResourceType) {
                            coapAddresses.forEach((address) => {
                                _sendCoapDiscovery(address, path);
                            });
                        }
                    });
                }
```

File: nodes/wot-discovery.html [hunk 3]
```diff
@@ -167,6 +169,30 @@
...
 
+    <div class="form-row" id="node-input-wellKnownURI-row">
+        <label for="node-input-wellKnownURI-row" style="width:auto"><i class="fa fa-tasks"></i> Well-known URI</label>
+        <div>
+            <input
+            type="checkbox"
+            id="node-input-tdURI"
+            style="display: inline-block; width: auto; vertical-align: top;"
+            />
+            <label for="node-input-useCoap" style="width: auto">
```
Comment (JKRhb) [inline nodes/wot-discovery.html:180]:
```suggestion
            <label for="node-input-tdURI" style="width: auto">
```

File: nodes/wot-discovery.html [hunk 4]
```diff
@@ -167,6 +169,30 @@
         </label>
     </div>
 
+    <div class="form-row" id="node-input-wellKnownURI-row">
+        <label for="node-input-wellKnownURI-row" style="width:auto"><i class="fa fa-tasks"></i> Well-known URI</label>
```
Comment (JKRhb) [inline nodes/wot-discovery.html:173]:
```suggestion
        <label for="node-input-wellKnownURI-row" style="width:auto"><i class="fa fa-tasks"></i> Well-known URIs</label>
```

File: nodes/wot-discovery.html [hunk 5]
```diff
@@ -167,6 +169,30 @@
...
+            </label>
+        </div>
+        <div>
+            <input
+            type="checkbox"
+            id="node-input-coreURI"
+            style="display: inline-block; width: auto; vertical-align: top;"
+            />
+            <label for="node-input-coreURI" style="width: auto">
+                Use core
```
Comment (JKRhb) [inline nodes/wot-discovery.html:191]: I haven't tested it yet but maybe we can wrap those URIs in `<code>` tags for better highlighting.

```suggestion
                Use <code>/.well-known/core</code>
```

File: nodes/wot-discovery.js [hunk 6]
```diff
@@ -22,6 +24,11 @@ module.exports = function (RED) {
             coapAddresses = _getCoapAddresses(config);
         }
 
+        if(coreURI){
+            if(config.useCoap){
+            }
+        }
+
```
Comment (JKRhb) [inline nodes/wot-discovery.js:31]: Are these lines still needed?

File: nodes/wot-discovery.js [hunk 7]
```diff
@@ -147,6 +159,39 @@ module.exports = function (RED) {
...
+                    reqOpt.query =`rt=wot.thing`;
+                    reqOpt.multicast = true;
+                    var req = coap.request(reqOpt);
+                    req.on("response", _coreResponse);
+                    req.on("error", function (err) {
+                        node.log("client error");
+                        node.log(err);
+                    });
+                    req.end();
+                });
```
Comment (JKRhb) [inline nodes/wot-discovery.js:176]: Just some minor code styling. 

(Not entirely sure about the `const`s, though, but it should work.)

```suggestion
            if (config.useCoap) {
                coapAddresses.forEach(address => {
                    const reqOpt s= url.parse(`coap://${address}/.well-known/core`);
                    reqOpts.pathname = reqOpts.path;
                    reqOpts.query ="rt=wot.thing";
                    reqOpts.multicast = true;
                    const req = coap.request(reqOpts);
                    req.on("response", _coreResponse);
                    req.on("error", function (err) {
                        node.log("client error");
                        node.log(err);
                    });
                    req.end();
                });
```

File: nodes/wot-discovery.js (hunk 6 above)
Comment (schulztr) [reply nodes/wot-discovery.js]: No they ares useless. I removed them in dff99200e621c94ca3d7ee4f954569419729f2c9

**Comment by JKRhb** [comment]
(Ehem, sorry for the linting error in my suggestion.)

File: nodes/wot-discovery.js [hunk 8]
```diff
@@ -147,6 +154,74 @@ module.exports = function (RED) {
...
+                                case "rt":
+                                    if (values === '"wot.thing"') {
+                                        correctResourceType = true;
+                                    }
+                                    break;
+                            }
+                        })
+
+                        if (correctContentType && correctResourceType) {
+                            _sendCoapDiscovery(`[${res.rsinfo.address}]`, path);
```
Comment (JKRhb) [inline nodes/wot-discovery.js:219]: One minor thing: In the [RFC](https://datatracker.ietf.org/doc/html/rfc6690#section-2) I noticed that the URI-Reference (here currently denoted as "path") can also be an absolute one. Therefore I think that we have to perform some parsing (probably using `urllib`) to determine if it is a relative or an absolute URI.

File: nodes/wot-discovery.js [hunk 9]
```diff
@@ -110,9 +117,9 @@ module.exports = function (RED) {
             return identifier;
         }
 
-        function _sendCoapDiscovery(address) {
+        function _sendCoapDiscovery(address, knownURI) {
```
Comment (JKRhb) [inline nodes/wot-discovery.js:120]:
```suggestion
        function _sendCoapDiscovery(address, path) {
```

File: nodes/wot-discovery.js [hunk 10]
```diff
@@ -110,9 +117,9 @@ module.exports = function (RED) {
...
             var reqOpts = url.parse(
-                `coap://${address}/.well-known/wot-thing-description`
+                `coap://${address}${knownURI}`
```
Comment (JKRhb) [inline nodes/wot-discovery.js:122]:
```suggestion
                `coap://${address}${path}`
```

File: nodes/wot-discovery.js (hunk 8 above)
Comment (schulztr) [reply nodes/wot-discovery.js:219]: I added a ToDo in the description for this.

**Review by JKRhb — CHANGES_REQUESTED** [review]
Hmm, I guess this should be changed to something more generic like "path" as the URI that it is obtained during CoRE resource discovery does not have to be a well-known one.

File: nodes/wot-discovery.js [hunk 11]
```diff
@@ -36,7 +38,12 @@ module.exports = function (RED) {
 
             if (coapAddresses) {
                 coapAddresses.forEach((address) => {
-                    _sendCoapDiscovery(address);
+                    if(tdURI){
+                        _sendCoapDiscovery(address, "/.well-known/wot-thing-description");
+                    }
+                    if(coreURI){
+                        _getDiscoveryLinks();
+                    }
```
Comment (JKRhb) [inline nodes/wot-discovery.js:46]: Ehem, and another very minor style fix.

```suggestion
                    if (tdURI) {
                        _sendCoapDiscovery(address, "/.well-known/wot-thing-description");
                    }
                    if (coreURI) {
                        _getDiscoveryLinks();
                    }
```

**Review by JKRhb — APPROVED** [review]

File: nodes/wot-discovery.js [hunk 12]
```diff
@@ -148,6 +155,78 @@ module.exports = function (RED) {
...
+                                    break;
+                            }
+                        });
+
+                        if (correctContentType && correctResourceType) {
+                            const uri = url.parse(path);
+                            if(uri.host){
+                                _sendCoapDiscovery(uri.host, uri.path);
+                            }else{
+                                _sendCoapDiscovery(`[${res.rsinfo.address}]`, uri.path);
```
Comment (JKRhb) [inline nodes/wot-discovery.js:223]:
```suggestion
                            if (uri.host) {
                                _sendCoapDiscovery(uri.host, uri.path);
                            } else {
                                _sendCoapDiscovery(`[${res.rsinfo.address}]`, uri.path);
```

**Pull request merged**